File: fotoapparat/src/test/java/io/fotoapparat/result/transformer/SaveToFileTransformerTest.java
package io.fotoapparat.result.transformer;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.File;

import io.fotoapparat.photo.Photo;
import io.fotoapparat.util.ExifOrientationWriter;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class SaveToFileTransformerTest {

    static final File FILE = new File("test");

    @Mock
    ExifOrientationWriter exifOrientationWriter;

    SaveToFileTransformer testee;

    @Before
    public void setUp() throws Exception {
        ensureFileDeleted();

        testee = new SaveToFileTransformer(
                FILE,
                exifOrientationWriter
        );
    }

    @Test
    public void savePhoto() throws Exception {
        // Given
        Photo photo = new Photo(
                new byte[]{1, 2, 3},
                0
        );

        // When
        testee.transform(photo);

        // Then
        assertTrue(FILE.exists());
        assertEquals(
                photo.encodedImage.length,
                FILE.length()
        );

        verify(exifOrientationWriter).writeExifOrientation(FILE, photo);
    }

    @After
    public void tearDown() throws Exception {
        ensureFileDeleted();
    }

    private void ensureFileDeleted() {
        if (FILE.exists() && !FILE.delete()) {
            throw new IllegalStateException("Can't delete test file");
        }
    }
}

File: fotoapparat/src/test/java/io/fotoapparat/result/transformer/SaveToFileTransformerTest.kt
package io.fotoapparat.result.transformer

import io.fotoapparat.exif.ExifOrientationWriter
import io.fotoapparat.result.Photo
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@RunWith(MockitoJUnitRunner::class)
internal class SaveToFileTransformerTest {

    val file = File("test")
    @Mock
    lateinit var exifOrientationWriter: ExifOrientationWriter
    lateinit var testee: SaveToFileTransformer

    @Before
    fun setUp() {
        file.ensureDeleted()

        testee = SaveToFileTransformer(
                file,
                exifOrientationWriter
        )
    }

    @Test
    fun `Save photo`() {
        // Given
        val photo = Photo(
                encodedImage = byteArrayOf(1, 2, 3),
                rotationDegrees = 0
        )

        // When
        testee(photo)

        // Then
        assertTrue(file.exists())
        assertEquals(
                expected = photo.encodedImage.size.toLong(),
                actual = file.length()
        )

        verify(exifOrientationWriter).writeExifOrientation(
                file,
                photo.rotationDegrees
        )
    }

    @After
    fun tearDown() {
        file.ensureDeleted()
    }

}

private fun File.ensureDeleted() {
    if (exists() && !delete()) {
        throw IllegalStateException("Can't delete test file")
    }
}
